shared/Microsoft.Extensions.ObjectMethodExecutor.Sources/ObjectMethodExecutor.cs

```diff
@@ -0,0 +1,453 @@
+                getResultMethod = null;
+                returnType = null;
+                return false;
+            }
+
+            awaiterType = getAwaiterMethod.ReturnType;
+
+            // Awaiter must have property matching "bool IsCompleted { get; }"
+            isCompletedProperty = awaiterType.GetRuntimeProperties().FirstOrDefault(p =>
+                p.Name.Equals("IsCompleted", StringComparison.OrdinalIgnoreCase)
+                && p.PropertyType == typeof(bool)
+                && p.GetMethod != null);
+            if (isCompletedProperty == null)
+            {
+                onCompletedMethod = null;
+                unsafeOnCompletedMethod = null;
+                getResultMethod = null;
+                returnType = null;
+                return false;
+            }
+
+            // Awaiter must implement INotifyCompletion
+            var implementsINotifyCompletion = awaiterType.GetInterfaces().Any(t => t == typeof(INotifyCompletion));
+            if (implementsINotifyCompletion)
+            {
+                // INotifyCompletion supplies a method matching "void OnCompleted(Action action)"
+                var interfaceMethods = awaiterType.GetInterfaceMap(typeof(INotifyCompletion)).InterfaceMethods;
+                onCompletedMethod = interfaceMethods.Single(m =>
+                    m.Name.Equals("OnCompleted", StringComparison.OrdinalIgnoreCase)
+                    && m.ReturnType == typeof(void)
+                    && m.GetParameters().Length == 1
+                    && m.GetParameters()[0].ParameterType == typeof(Action));
+            }
+            else
+            {
+                onCompletedMethod = null;
+                unsafeOnCompletedMethod = null;
+                getResultMethod = null;
+                returnType = null;
+                return false;
+            }
+
+            // Awaiter optionally implements ICriticalNotifyCompletion
+            var implementsICriticalNotifyCompletion = awaiterType.GetInterfaces().Any(t => t == typeof(ICriticalNotifyCompletion));
+            if (implementsICriticalNotifyCompletion)
+            {
+                // ICriticalNotifyCompletion supplies a method matching "void UnsafeOnCompleted(Action action)"
+                var interfaceMethods = awaiterType.GetInterfaceMap(typeof(ICriticalNotifyCompletion)).InterfaceMethods;
+                unsafeOnCompletedMethod = interfaceMethods.Single(m =>
+                    m.Name.Equals("UnsafeOnCompleted", StringComparison.OrdinalIgnoreCase)
+                    && m.ReturnType == typeof(void)
+                    && m.GetParameters().Length == 1
+                    && m.GetParameters()[0].ParameterType == typeof(Action));
+            }
+            else
+            {
+                unsafeOnCompletedMethod = null;
+            }
+
+            // Awaiter must have method matching "void GetResult" or "T GetResult()"
+            getResultMethod = awaiterType.GetRuntimeMethods().FirstOrDefault(m =>
```

Comment:
The two calls to `GetRuntimeMethods` are on different types (the first one is on `type` (the thing we're checking whether is awaitable); the second one is on `awaiterType`).